futures-util/src/try_stream/try_filter.rs

```diff
@@ -0,0 +1,118 @@
+use core::pin::Pin;
+use futures_core::future::Future;
+use futures_core::stream::{Stream, TryStream};
+use futures_core::task::{Context, Poll};
+use futures_sink::Sink;
+use pin_utils::{unsafe_pinned, unsafe_unpinned};
+
+/// Stream for the [`try_filter`](super::TryStreamExt::try_filter)
+/// method.
+#[derive(Debug)]
+#[must_use = "streams do nothing unless polled"]
+pub struct TryFilter<St, Fut, F>
+    where St: TryStream
+{
+    stream: St,
+    f: F,
+    pending_fut: Option<Fut>,
+    pending_item: Option<St::Ok>,
+}
+
+impl<St, Fut, F> Unpin for TryFilter<St, Fut, F>
+    where St: TryStream + Unpin, Fut: Unpin,
+{}
+
+impl<St, Fut, F> TryFilter<St, Fut, F>
+    where St: TryStream
+{
+    unsafe_pinned!(stream: St);
+    unsafe_unpinned!(f: F);
+    unsafe_pinned!(pending_fut: Option<Fut>);
+    unsafe_unpinned!(pending_item: Option<St::Ok>);
+
+    pub(super) fn new(stream: St, f: F) -> Self {
+        TryFilter {
+            stream,
+            f,
+            pending_fut: None,
+            pending_item: None,
+        }
+    }
+
+    /// Acquires a reference to the underlying stream that this combinator is
+    /// pulling from.
+    pub fn get_ref(&self) -> &St {
+        &self.stream
+    }
+
+    /// Acquires a mutable reference to the underlying stream that this
+    /// combinator is pulling from.
+    ///
+    /// Note that care must be taken to avoid tampering with the state of the
+    /// stream which may otherwise confuse this combinator.
+    pub fn get_mut(&mut self) -> &mut St {
+        &mut self.stream
+    }
+
+    /// Acquires a pinned mutable reference to the underlying stream that this
+    /// combinator is pulling from.
+    ///
+    /// Note that care must be taken to avoid tampering with the state of the
+    /// stream which may otherwise confuse this combinator.
+    pub fn get_pin_mut<'a>(self: Pin<&'a mut Self>) -> Pin<&'a mut St> {
+        unsafe { Pin::map_unchecked_mut(self, Self::get_mut) }
+    }
+
+    /// Consumes this combinator, returning the underlying stream.
+    ///
+    /// Note that this may discard intermediate state of this combinator, so
+    /// care should be taken to avoid losing resources when this is called.
+    pub fn into_inner(self) -> St {
+        self.stream
+    }
+}
+
+impl<St, Fut, F> Stream for TryFilter<St, Fut, F>
+    where St: TryStream,
+          Fut: Future<Output = bool>,
+          F: FnMut(&St::Ok) -> Fut,
+{
+    type Item = Result<St::Ok, St::Error>;
+
+    fn poll_next(
+        mut self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+    ) -> Poll<Option<Result<St::Ok, St::Error>>> {
+        loop {
+            if self.as_mut().pending_fut().as_pin_mut().is_none() {
```

Comment:
```suggestion
            if self.pending_fut.is_none() {
```
It is unnecessary to do pin projection here because `Pin` implements `Deref` unconditionally.

Reply:
I was using the same implementation as a [Stream::filter](https://github.com/rust-lang-nursery/futures-rs/blob/5a6b1b32329b3f4a0ad1df3df1576fb36fdc5361/futures-util/src/stream/filter.rs#L102).

This line can be reduced down to this:

```rust
if self.as_mut().pending_fut().is_none()
```
, but the change you suggested is failing with the following error:
```rust
error[E0382]: use of moved value: `self`
  --> futures-util/src/try_stream/try_filter.rs:87:16
   |
83 |         mut self: Pin<&mut Self>,
   |         -------- move occurs because `self` has type `std::pin::Pin<&mut try_stream::try_filter::TryFilter<St, Fut, F>>`, which does not implement the `Copy` trait
...
87 |             if self.pending_fut().is_none() {
   |                ^^^^ value moved here, in previous iteration of loop

```

Reply:
Seems to be calling a method. It needs to remove "`()`" like this.
```rust
if self.pending_fut.is_none() {
```

Reply:
My bad, fixed.